### ISSUE TITLE
ReferenceError: require is not defined

### ISSUE THREAD

**Issue by martinszeltins** [electron-vite/vite-plugin-electron-renderer#24]
I am getting this error when I added Vue devtools:

> Uncaught ReferenceError: require is not defined at events.js?v=e1f2c7ab:1:13

Here is my main.ts

````
import devtools from '@vue/devtools'
import App from '@/App.vue'
import pinia from '@/stores'
import router from '@/routes'
import { createApp } from 'vue'

devtools.connect()

const app = createApp(App)

app.use(pinia)
    .use(router)
    .mount('#app')
````

And here is my vite config:

````
import path from 'path'
import pkg from './package.json'
import { rmSync } from 'node:fs'
import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'
import { fileURLToPath, URL } from 'url'
import electron from 'vite-plugin-electron'
import Components from 'unplugin-vue-components/vite'
import electronRenderer from 'vite-plugin-electron-renderer'

rmSync('dist-electron', { recursive: true, force: true })

const isProduction = process.env.NODE_ENV === 'production'
const isDevelopment = process.env.NODE_ENV === 'development' || !!process.env.VSCODE_DEBUG

export default defineConfig({
    plugins: [
        vue({ reactivityTransform: true }),
        
        Components(),

        electron([
            {
                entry: 'electron/main.ts',
                onstart(options) {
                    if (process.env.VSCODE_DEBUG) {
                        /* eslint-disable no-console */
                        console.log('[startup] Electron App')
                    } else {
                        options.startup()
                    }
                },
                
                vite: {
                    build: {
                        sourcemap: isDevelopment,
                        minify: isProduction,
                        outDir: 'dist-electron/main',
                        rollupOptions: {
                            external: Object.keys('dependencies' in pkg ? pkg.dependencies : {}),
                        },
                    },
                },
            },

            {
                entry: 'electron/preload.ts',
                onstart(options) {
                    options.reload()
                },
                vite: {
                    build: {
                        sourcemap: isDevelopment,
                        minify: isProduction,
                        outDir: 'dist-electron/preload',
                        rollupOptions: {
                            external: Object.keys('dependencies' in pkg ? pkg.dependencies : {}),
                        },
                    },
                },
            }
        ]),

        electronRenderer({
            nodeIntegration: false,
        }),
    ],

    resolve: {
        alias: {
            // @ts-ignore
            '@': fileURLToPath(new URL('./src', import.meta.url)),
            '@img': fileURLToPath(new URL('./src/assets/img', import.meta.url)),
        },
    },

    /* eslint-disable no-extra-boolean-cast */
    server: !!process.env.VSCODE_DEBUG ? (() => {
        const url = new URL(pkg.debug.env.VITE_DEV_SERVER_URL)
        return {
            host: url.hostname,
            port: +url.port,
        }
    })() : undefined,

    clearScreen: false,

    build: {
        sourcemap: true,
        rollupOptions: {
            input: {
                index: path.join(__dirname, 'index.html'),
            },
        },
    },
})
````

And here is my package.json with all versions:

````
{
    "name": "electron-vue-vite",
    "version": "2.0.0",
    "main": "dist-electron/main/main.js",
    "description": "Really simple Electron + Vue + Vite boilerplate.",
    "author": "Martins Zeltins <martins@martinsz.lv>",
    "license": "MIT",
    "private": true,
    "debug": {
        "env": {
            "VITE_DEV_SERVER_URL": "http://127.0.0.1:3344/"
        }
    },
    "scripts": {
        "dev": "vite",
        "build": "npm run check && vite build && electron .",
        "typecheck": "vue-tsc --noEmit && tsc --noEmit ./electron/*.ts && tsc --noEmit ./electron/**/*.ts",
        "lint": "eslint . --ext .vue,.js,.ts,.tsx --ignore-path .gitignore --max-warnings 0",
        "check": "npm run typecheck && npm run lint",
        "postinstall": "electron-builder install-app-deps",
        "devtools": "./node_modules/.bin/vue-devtools"
    },
    "devDependencies": {
        "@vitejs/plugin-vue": "4.0.0",
        "electron": "22.0.2",
        "electron-builder": "23.6.0",
        "typescript": "4.9.4",
        "vite": "4.0.4",
        "vite-plugin-electron": "0.11.1",
        "vite-plugin-electron-renderer": "0.11.4",
        "vue": "3.2.45",
        "vue-tsc": "1.0.24",
        "@vue/eslint-config-typescript": "11.0.0",
        "eslint": "8.20.0",
        "eslint-plugin-vue": "9.2.0",
        "vue-eslint-parser": "9.0.3",
        "unplugin-vue-components": "0.22.12"
    },
    "dependencies": {
        "vue-router": "4.1.6",
        "sqlite": "4.1.2",
        "sqlite3": "5.1.4",
        "nock": "13.3.0",
        "mock-aws-s3": "4.0.2",
        "aws-sdk": "2.1295.0",
        "@typescript-eslint/parser": "5.30.7",
        "tailwindcss": "3.2.4",
        "postcss": "8.4.21",
        "autoprefixer": "10.4.13",
        "postcss-important-startstop": "2.0.0",
        "@tailwindcss/container-queries": "0.1.0",
        "postcss-import": "15.1.0",
        "postcss-nested": "6.0.0",
        "@tailwindcss/nesting": "0.0.0-insiders.565cd3e",
        "pinia": "2.0.29",
        "primevue": "3.22.1",
        "primeicons": "6.0.1",
        "radash": "10.7.0",
        "mitt": "3.0.0",
        "@vue/devtools": "6.5.0"
    }
}
````

**Comment by caoxiemeihao** [comment]
Ensure that `nodeIntegration` is enabled in the Main Process.

**Comment by martinszeltins** [comment]
> Ensure that `nodeIntegration` is enabled in the Main Process.

`nodeIntegration` is disabled in my app for security purposes.

**Comment by caoxiemeihao** [comment]
Can you provide a minimalist replica of the repo?

**Comment by rzg2022** [comment]
When I set `nodeIntegration` to **false** in the main process, the program starts with an **`Uncaught ReferenceError: require is not defined`**. How do I resolve this?

```
// vite.config.js
renderer({
 nodeIntegration: false
})
```

```
// index.ts in main
win = new BrowserWindow({
  title: 'Main window',
  icon: join(process.env.PUBLIC, 'favicon.ico'),
  webPreferences: {
    preload,
    // Warning: Enable nodeIntegration and disable contextIsolation is not secure in production
    // Consider using contextBridge.exposeInMainWorld
    // Read more on https://www.electronjs.org/docs/latest/tutorial/context-isolation
    nodeIntegration: false,
    contextIsolation: true
  }
})
```

**Comment by martinszeltins** [comment]
Using Electron and Node.js API in Renderer process is not safe or recommended.

> renderer/index.ts
> 
> ```diff
> - import './samples'
> 
> document.getElementById('app')!.innerHTML = `
> <h1>Hi there 👋</h1>
> <p>Now, you can use Electron and Node.js API in Renderer process.</p>
> <pre>
>   import { ipcRenderer } from 'electron'
>   import fs from 'fs'
> </pre>
> `
> ```

**Comment by caoxiemeihao** [comment]
Everyone has everyone's decision.

**Comment by MarMun** [comment]
Same problem here. Tried everything but to no avail.

I'm about to just move every module dependent of a node package to the main process (and connected via contextBridge). 

Question: Can I remove renderer plugin if I don't need any node polyfills or is there some magic performed I still need?

**Comment by caoxiemeihao** [comment]
@MarMun Can you  provide a minimal reproduction repo?

**Comment by caoxiemeihao** [comment]
@martinszeltins Hey! 👋
Maybe [vite-plugin-target](https://github.com/vite-plugin/vite-plugin-target/) is better for you. :)

**Comment by MarMun** [comment]
@caoxiemeihao here you go: [fork](https://github.com/MarMun/electron-vite-vue_require-not-defined).

I also need nodeIntegration disabled along with other recommended config:

```
webSecurity: true,
// https://www.electronjs.org/docs/all#2-do-not-enable-nodejs-integration-for-remote-content
nodeIntegration: false,
// https://www.electronjs.org/docs/all#context-isolation
contextIsolation: true,
// https://www.electronjs.org/docs/all#15-disable-the-remote-module
enableRemoteModule: false,
sandbox: true
```
Thx for the boilerplate and your help!

**Comment by caoxiemeihao** [comment]
@MarMun 
https://github.com/MarMun/electron-vite-vue_require-not-defined/blob/main/src/bufferTest.js
<img width="375" alt="image" src="https://user-images.githubusercontent.com/26263658/217414144-d4c7a622-c8f8-4a75-8756-3f7b5d446c4a.png">

`buffer` is Node.js native module. 
You don't want to enable `nodeIntegration` in the Renderer, but you want to use the Node.js native module! What do you want?

**Comment by MarMun** [comment]
> You don't want to enable nodeIntegration in the Renderer, but you want to use the Node.js native module! What do you want?

The 'buffer' node module (or anything else I may want to polyfill) doesn't need to be executed in / needs access to main (node) context. It can / should run in renderer scope / context.

I may be wrong but aren't  node module polyfills designed for the browser which has no node context access?

The problem is, I guess, that (`cjs`) modules polyfilled are not  converted to `ES6` (and we want a  `ESNext` app as is configured in `tsconfig.json` / `tsconfig.node.json`).

I tried to do this manually with plugins like
- [@rollup/plugin-commonjs](https://www.npmjs.com/package/@rollup/plugin-commonjs) for production and 
- [vite-plugin-commonjs](https://github.com/vite-plugin/vite-plugin-commonjs) for development

but I failed to make it work (it always breaks with `ReferenceError: require is not defined`).
